### DISCUSSION TITLE
Helper class classloading issue

### DISCUSSION THREAD

**Discussion by kamilharksealights** [open-telemetry/opentelemetry-java-instrumentation#5607] [Q&A]
Hi,
I have an agent extension providing InstrumentationModule that uses a helper class (listed in getAdditionalHelperClassNames)
While transforming the code I use this helper class to register some state, let's say I want to store all methods I transformed. 
I also inject some bytecode to the methods with usage of this helper class.
The issue is that this helper class is loaded twice by two classloaders
1. while transforming the code, it's io.opentelemetry.javaagent.tooling.ExtensionClassLoader
2. at runtime it's jdk.internal.loader.ClassLoaders$AppClassLoader
and because of that the state I collect during transformation is not available at runtime
I guess, it would all work if my class was loaded by bootstrap classloader, by I have no clue how to do that
(I use gradle to produce the final agent jar, with extensions folder with my code inside)

**Comment by trask** [comment]
hi @kamilharksealights!

> while transforming the code, it's io.opentelemetry.javaagent.tooling.ExtensionClassLoader

I don't _think_ the helper classes are normally initialized in the agent (extension) class loader. Maybe you are referencing the class directly from your InstrumentationModule?

> because of that the state I collect during transformation is not available at runtime

what kind of state do you want to collect during transformation? can that be collected in the static initializer of your helper class instead?

**Comment by laurit** [comment]
@kamilharksealights what you have observed is the expected behaviour. The classes that are used during transformation are in a different class loader (in agent or extension class loader) from the helpers that are used from the transformed bytecode (are in the same class loader as the transformed class). If you are already building your own agent jar you can just add classes to the agent jar, these will be available in boot loader and you can use them to exchange information between transformation and runtime code. You may need to use a `BootstrapPackagesConfigurer` to let the agent know which package contains your classes.

**Reply by kamilharksealights** [reply]
@laurit, this is what I did in the end and it works nice. thx!
another question, a bit unrelated, is there any place/hook I could use to run my code after all the instrumentation/transformation is done?

**Reply by kamilharksealights** [reply]
hmm, I think I know the answer - there is no such place because any class can be loaded `later` and it will be transformed then, correct?

**Reply by laurit** [reply]
@kamilharksealights you are correct, classes are loaded as needed and transformed when the class is defined.
